include/nbl/system/ICancelableAsyncQueueDispatcher.h

```diff
@@ -0,0 +1,179 @@
+#ifndef __NBL_I_CANCELABLE_ASYNC_QUEUE_DISPATCHER_H_INCLUDED__
+#define __NBL_I_CANCELABLE_ASYNC_QUEUE_DISPATCHER_H_INCLUDED__
+
+#include "nbl/system/IAsyncQueueDispatcher.h"
+#include "nbl/system/SReadWriteSpinLock.h"
+
+namespace nbl {
+namespace system
+{
+
+namespace impl
+{
+    class ICancelableAsyncQueueDispatcherBase
+    {
+    public:
+        class future_base_t;
+
+        struct request_base_t : impl::IAsyncQueueDispatcherBase::request_base_t
+        {
+            //! Atomically cancels this request
+            bool set_cancel();
+            bool query_cancel() const
+            {
+                return future == nullptr;
+            }
+
+        private:
+            friend future_base_t;
+            friend ICancelableAsyncQueueDispatcherBase;
+
+            //! See ICancelableAsyncQueueDispatcher::associate_request_with_future() docs
+            void associate_future_object(future_base_t* _future)
+            {
+                future = _future;
+            }
+
+            future_base_t* future = nullptr;
+        };
+
+        class future_base_t
+        {
+            friend request_base_t;
+
+        protected:
+            std::atomic_bool valid_flag = false;
+            std::atomic<request_base_t*> request = nullptr;
+
+            // future_t is non-copyable and non-movable
+            future_base_t(const future_base_t&) = delete;
+
+        public:
+            future_base_t() = default;
+            ~future_base_t()
+            {
+                if (request)
+                    request.load()->set_cancel();
```

Comment:
you need to impement this via a `future_base_t::cancel()`

and this need to work like this
```cpp
auto hadARequest = request.exchange(nullptr);
if (hadARequest)
    hadARequest->set_skip(); // call it abort/cancel
request = nullptr;
```

because of the race condition between the thread where the future's lifetime is defined and the worker thread

Reply:
conversation moved.